pkg/envoy/cds/builder.go

```diff
@@ -28,9 +31,139 @@ import (
 	"github.com/openservicemesh/osm/pkg/trafficpolicy"
 )
 
+type clusterBuilder struct {
+	proxyIdentity identity.ServiceIdentity
+
+	outboundMeshTrafficPolicy *trafficpolicy.OutboundMeshTrafficPolicy
+
+	inboundMeshTrafficPolicy *trafficpolicy.InboundMeshTrafficPolicy
+
+	sidecarSpec configv1alpha2.SidecarSpec
+
+	egressTrafficPolicy *trafficpolicy.EgressTrafficPolicy
+
+	envoyOutboundPassthroughCluster string
+
+	egressEnabled bool
+
+	metricsEnabled bool
+
+	tracingEnabled bool
+
+	tracingAddress string
+
+	tracingPort uint32
+}
+
+func NewClusterBuilder() *clusterBuilder {
+	return &clusterBuilder{}
+}
+
+func (b *clusterBuilder) SetProxyIdentity(identity identity.ServiceIdentity) *clusterBuilder {
+	b.proxyIdentity = identity
+	return b
+}
+
+func (b *clusterBuilder) SetOutboundMeshTrafficPolicy(outboundTrafficPolicy *trafficpolicy.OutboundMeshTrafficPolicy) *clusterBuilder {
+	b.outboundMeshTrafficPolicy = outboundTrafficPolicy
+	return b
+}
+
+func (b *clusterBuilder) SetSidecarSpec(spec configv1alpha2.SidecarSpec) *clusterBuilder {
+	b.sidecarSpec = spec
+	return b
+}
+
+func (b *clusterBuilder) SetProxyServices(proxyServices []service.MeshService) *clusterBuilder {
+	b.sidecarSpec = spec
+	return b
+}
+
+func (b *clusterBuilder) SetInboundMeshTrafficPolicy(inboundTrafficPolicy *trafficpolicy.InboundMeshTrafficPolicy) *clusterBuilder {
+	b.inboundMeshTrafficPolicy = inboundTrafficPolicy
+	return b
+}
+
+func (b *clusterBuilder) SetEgressTrafficPolicy(egressTrafficPolicy *trafficpolicy.EgressTrafficPolicy) *clusterBuilder {
+	b.egressTrafficPolicy = egressTrafficPolicy
+	return b
+}
+
+func (b *clusterBuilder) SetEnvoyOutBoundPassthroughCluster(envoyOutboundPassthroughCluster string) *clusterBuilder {
+	b.envoyOutboundPassthroughCluster = envoyOutboundPassthroughCluster
+	return b
+}
+
+func (b *clusterBuilder) SetEgressEnabled(egressEnabled bool) *clusterBuilder {
+	b.egressEnabled = egressEnabled
+	return b
+}
+
+func (b *clusterBuilder) SetMetricsEnabled(metricsEnabled bool) *clusterBuilder {
+	b.metricsEnabled = metricsEnabled
+	return b
+}
+
+func (b *clusterBuilder) SetTracingEnabled(tracingEnabled bool) *clusterBuilder {
+	b.tracingEnabled = tracingEnabled
+	return b
+}
+
+func (b *clusterBuilder) SetTracingAddress(tracingAddress string) *clusterBuilder {
+	b.tracingAddress = tracingAddress
+	return b
+}
+
+func (b *clusterBuilder) SetTracingPort(tracingPort uint32) *clusterBuilder {
+	b.tracingPort = tracingPort
+	return b
+}
+
 // replacer used to configure an Envoy cluster's altStatName
 var replacer = strings.NewReplacer(".", "_", ":", "_")
 
+func (b *clusterBuilder) Build() ([]*xds_cluster.Cluster, error) {
+	var clusters []*xds_cluster.Cluster
+
+	// Build upstream clusters based on allowed outbound traffic policies
+	if b.outboundMeshTrafficPolicy != nil {
+		clusters = append(clusters, upstreamClustersFromClusterConfigs(b.proxyIdentity, b.outboundMeshTrafficPolicy.ClustersConfigs, b.sidecarSpec)...)
```

Comment:
instead of passing these fields in, I'd consider making `upstreamClustersFromClusterConfigs` a method on the builder which has access to those fields. ie:

```go
clusters = append(clusters, b.buildUpstreamClusters()
```